libraries/chain/controller.cpp

```diff
@@ -1424,18 +1456,35 @@ struct controller_impl {
                      ("lib_num", lib_num)("bn", fork_db_root_block_num()) );
       }
 
-      // maintain legacy only advancing LIB via validated blocks, hence pass in chain_head id for use
-      const block_id_type new_lib_id = fork_db.pending_lib_id(irreversible_mode() ? block_id_type{} : chain_head.id());
+      auto pending_lib_id = [&]() {
+         return fork_db.apply<block_id_type>(
+            [&](const fork_database_legacy_t& forkdb) -> block_id_type {
+               // maintain legacy only advancing LIB via validated blocks, hence pass in chain_head id for use
+               block_state_legacy_ptr head = irreversible_mode() ? forkdb.head() : forkdb.get_block(chain_head.id());
+               if (!head)
+                  return {};
+               block_num_type dpos_lib_num = head->irreversible_blocknum();
+               block_state_legacy_ptr lib = forkdb.search_on_branch(head->id(), dpos_lib_num, include_root_t::no);
+               if (!lib)
+                  return {};
+               return lib->id();
+            },
+            [&](const fork_database_if_t& forkdb) -> block_id_type {
+               return forkdb.pending_savanna_lib_id();
+            }
+         );
+      };
+
+      const block_id_type new_lib_id = pending_lib_id();
       const block_num_type new_lib_num = block_header::num_from_id(new_lib_id);
 
       if( new_lib_num <= lib_num )
          return;
 
-      bool savanna_transition_required = false;
       auto mark_branch_irreversible = [&, this](auto& forkdb) {
          assert(!irreversible_mode() || forkdb.head());
          const auto& head_id = irreversible_mode() ? forkdb.head()->id() : chain_head.id();
-         auto branch = forkdb.fetch_branch( head_id, new_lib_id);
+         auto branch = forkdb.fetch_branch( head_id, new_lib_id); // verifies lib is on head branch
```

Comment:
When in Savanna mode, should we mark_irreversible only if `new_lib_id` is an ancestor of `head_id`? If so, why?

Reply:
Added additional comments

Reply:
Thanks. I feel that at this point, we know that `head_id` is on the wrong fork (since `lib_id` is *not* an ancestor), so it seems that we would want to take action now, rather than delaying to when `maybe_switch_fork` is called, however I'm not advocation for a change now since this is delicate code. 

I can't help but feel this code could be rearranged in a way that makes it simpler and easier to follow (at least for me :-).